crypto/src/hash/rescue/rp64_256_jive/mod.rs

```diff
@@ -0,0 +1,751 @@
+        //   and set the last state element to 5 (the number of elements to be hashed).
+        // - if the value doesn't fit into a single field element, split it into two field
+        //   elements, copy them into state elements 5 and 6, and set the last state element
+        //   to 6.
+        let mut state = [BaseElement::ZERO; STATE_WIDTH];
+        state[INPUT1_RANGE].copy_from_slice(seed.as_elements());
+        state[INPUT2_RANGE.start] = BaseElement::new(value);
+        if value < BaseElement::MODULUS {
+            state[INPUT2_RANGE.end - 1] = BaseElement::new(DIGEST_SIZE as u64 + 1);
+        } else {
+            state[INPUT2_RANGE.start + 1] = BaseElement::new(value / BaseElement::MODULUS);
+            state[INPUT2_RANGE.end - 1] = BaseElement::new(DIGEST_SIZE as u64 + 2);
+        }
+
+        let initial_state = state;
+        // apply the Rescue permutation and apply the final Jive summation
+        Self::apply_permutation(&mut state);
+
+        apply_jive_summation(&initial_state, &state)
+    }
+}
+
+impl ElementHasher for RpJive64_256 {
+    type BaseField = BaseElement;
+
+    fn hash_elements<E: FieldElement<BaseField = Self::BaseField>>(elements: &[E]) -> Self::Digest {
+        // convert the elements into a list of base field elements
+        let elements = E::as_base_elements(elements);
+
+        // initialize state to all zeros, except for the last element of the capacity part, which
+        // is set to the number of elements to be hashed. this is done so that adding zero elements
+        // at the end of the list always results in a different hash.
+        let mut state = [BaseElement::ZERO; STATE_WIDTH];
+        state[CAPACITY_RANGE.start] = BaseElement::new(elements.len() as u64);
+
+        // absorb elements into the state one by one until the rate portion of the state is filled
+        // up; then apply the Rescue permutation and start absorbing again; repeat until all
+        // elements have been absorbed
+        let mut i = 0;
+        for &element in elements.iter() {
+            state[RATE_RANGE.start + i] += element;
+            i += 1;
+            if i % RATE_WIDTH == 0 {
+                Self::apply_permutation(&mut state);
+                i = 0;
+            }
+        }
+
+        // if we absorbed some elements but didn't apply a permutation to them (would happen when
+        // the number of elements is not a multiple of RATE_WIDTH), apply the Rescue permutation.
+        // we don't need to apply any extra padding because we injected total number of elements
+        // in the input list into the capacity portion of the state during initialization.
+        if i > 0 {
+            Self::apply_permutation(&mut state);
+        }
+
+        // return the first 4 elements of the state as hash result
+        ElementDigest::new(state[DIGEST_RANGE].try_into().unwrap())
+    }
+}
+
+#[inline(always)]
+fn apply_jive_summation(
```

Comment:
I wonder if this should go into `RpJive64_256` and be a public function (similar to `apply_round()`). Might be useful for building execution traces and maybe for defining AIR constraints as well.